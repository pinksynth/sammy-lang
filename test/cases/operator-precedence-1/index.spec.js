const {
  performSnapshotAssertions,
  expectCompiledEval,
} = require("../../helpers")
performSnapshotAssertions("Operator precedence", __dirname, {
  jsGlobals: "bcdeghijklmnpqrsxyz".split(""),
})

test("Operator precedence implementation", () => {
  expectCompiledEval(`2 * 3 + 4 / 5`).toEqual(6.8)
  expectCompiledEval(`(-10) + 3 % 5`).toEqual(-7)
  expectCompiledEval(`6 ^ 2 * 3`).toEqual(108)
  expectCompiledEval(`
		foo = %[bar:  @{3}]
		baz = %[quux: @{4}]
	
		foo.bar() * baz.quux()
	`).toEqual(12)
})
