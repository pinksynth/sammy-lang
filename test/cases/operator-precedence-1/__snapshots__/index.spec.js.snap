// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“¸ Operator precedence 1`] = `
[
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "NUMBER",
    "value": "2",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 4,
    "columnNumberStart": 3,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 4,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "NUMBER",
    "value": "3",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "OPERATOR_INFIX",
    "value": "+",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "NUMBER",
    "value": "4",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "OPERATOR_INFIX",
    "value": "/",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 13,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "NUMBER",
    "value": "5",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 14,
    "lineNumberEnd": 2,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "HYPHEN",
    "value": "-",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 3,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "NUMBER",
    "value": "10",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "OPERATOR_INFIX",
    "value": "+",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "NUMBER",
    "value": "3",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "OPERATOR_INFIX",
    "value": "%",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 13,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "NUMBER",
    "value": "5",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 14,
    "lineNumberEnd": 3,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "NUMBER",
    "value": "6",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 4,
    "columnNumberStart": 3,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "OPERATOR_INFIX",
    "value": "^",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 4,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "NUMBER",
    "value": "2",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "NUMBER",
    "value": "3",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 10,
    "lineNumberEnd": 4,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "VAR",
    "value": "x",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 3,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "FORWARD_PIPE",
    "value": "->",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "VAR",
    "value": "y",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "VAR",
    "value": "z",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 4,
    "lineNumberStart": 4,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 10,
    "lineNumberEnd": 5,
    "lineNumberStart": 4,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 4,
    "columnNumberStart": 3,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 4,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "VAR",
    "value": "b",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 7,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "COMPARE",
    "value": "==",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "VAR",
    "value": "c",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 12,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "OPERATOR_INFIX",
    "value": "||",
  },
  {
    "columnNumberEnd": 15,
    "columnNumberStart": 14,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 16,
    "columnNumberStart": 15,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "VAR",
    "value": "d",
  },
  {
    "columnNumberEnd": 17,
    "columnNumberStart": 16,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 17,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "COMPARE",
    "value": "!=",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 5,
    "lineNumberStart": 5,
    "tokenType": "VAR",
    "value": "e",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 21,
    "lineNumberEnd": 6,
    "lineNumberStart": 5,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "f",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 4,
    "columnNumberStart": 3,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 4,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "g",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 7,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "COMPARE",
    "value": ">=",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "h",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 12,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "OPERATOR_INFIX",
    "value": "&&",
  },
  {
    "columnNumberEnd": 15,
    "columnNumberStart": 14,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 16,
    "columnNumberStart": 15,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "i",
  },
  {
    "columnNumberEnd": 17,
    "columnNumberStart": 16,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 17,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "COMPARE",
    "value": "<=",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "j",
  },
  {
    "columnNumberEnd": 22,
    "columnNumberStart": 21,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 24,
    "columnNumberStart": 22,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "OPERATOR_INFIX",
    "value": "||",
  },
  {
    "columnNumberEnd": 25,
    "columnNumberStart": 24,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 26,
    "columnNumberStart": 25,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "k",
  },
  {
    "columnNumberEnd": 27,
    "columnNumberStart": 26,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 28,
    "columnNumberStart": 27,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "COMPARE",
    "value": ">",
  },
  {
    "columnNumberEnd": 29,
    "columnNumberStart": 28,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 30,
    "columnNumberStart": 29,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "l",
  },
  {
    "columnNumberEnd": 31,
    "columnNumberStart": 30,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 33,
    "columnNumberStart": 31,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "OPERATOR_INFIX",
    "value": "&&",
  },
  {
    "columnNumberEnd": 34,
    "columnNumberStart": 33,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 35,
    "columnNumberStart": 34,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "m",
  },
  {
    "columnNumberEnd": 36,
    "columnNumberStart": 35,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 37,
    "columnNumberStart": 36,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "COMPARE",
    "value": "<",
  },
  {
    "columnNumberEnd": 38,
    "columnNumberStart": 37,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 39,
    "columnNumberStart": 38,
    "lineNumberEnd": 6,
    "lineNumberStart": 6,
    "tokenType": "VAR",
    "value": "n",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 39,
    "lineNumberEnd": 7,
    "lineNumberStart": 6,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 2,
    "columnNumberStart": 1,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "VAR",
    "value": "o",
  },
  {
    "columnNumberEnd": 3,
    "columnNumberStart": 2,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 4,
    "columnNumberStart": 3,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 5,
    "columnNumberStart": 4,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 6,
    "columnNumberStart": 5,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "VAR",
    "value": "p",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 6,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "DOT",
    "value": ".",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "VAR",
    "value": "q",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "VAR",
    "value": "r",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "DOT",
    "value": ".",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 13,
    "lineNumberEnd": 7,
    "lineNumberStart": 7,
    "tokenType": "VAR",
    "value": "s",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 14,
    "lineNumberEnd": 8,
    "lineNumberStart": 7,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
]
`;

exports[`ðŸ“¸ Operator precedence 2`] = `
{
  "children": [
    {
      "columnNumberStart": 7,
      "left": {
        "columnNumberStart": 3,
        "left": {
          "columnNumberStart": 1,
          "lineNumberStart": 1,
          "type": "LITERAL_NUMBER",
          "value": "2",
        },
        "lineNumberStart": 1,
        "operator": "*",
        "right": {
          "columnNumberStart": 5,
          "lineNumberStart": 1,
          "type": "LITERAL_NUMBER",
          "value": "3",
        },
        "type": "BINARY_EXPR",
      },
      "lineNumberStart": 1,
      "operator": "+",
      "right": {
        "columnNumberStart": 11,
        "left": {
          "columnNumberStart": 9,
          "lineNumberStart": 1,
          "type": "LITERAL_NUMBER",
          "value": "4",
        },
        "lineNumberStart": 1,
        "operator": "/",
        "right": {
          "columnNumberStart": 13,
          "lineNumberStart": 1,
          "type": "LITERAL_NUMBER",
          "value": "5",
        },
        "type": "BINARY_EXPR",
      },
      "type": "BINARY_EXPR",
    },
    {
      "columnNumberStart": 7,
      "left": {
        "children": [
          {
            "columnNumberStart": 2,
            "lineNumberStart": 2,
            "operand": {
              "columnNumberStart": 3,
              "lineNumberStart": 2,
              "type": "LITERAL_NUMBER",
              "value": "10",
            },
            "operator": "-",
            "type": "UNARY_EXPRESSION",
          },
        ],
        "columnNumberStart": 1,
        "lineNumberStart": 2,
        "type": "GENERIC_EXPRESSION",
      },
      "lineNumberStart": 2,
      "operator": "+",
      "right": {
        "columnNumberStart": 11,
        "left": {
          "columnNumberStart": 9,
          "lineNumberStart": 2,
          "type": "LITERAL_NUMBER",
          "value": "3",
        },
        "lineNumberStart": 2,
        "operator": "%",
        "right": {
          "columnNumberStart": 13,
          "lineNumberStart": 2,
          "type": "LITERAL_NUMBER",
          "value": "5",
        },
        "type": "BINARY_EXPR",
      },
      "type": "BINARY_EXPR",
    },
    {
      "columnNumberStart": 7,
      "left": {
        "columnNumberStart": 3,
        "left": {
          "columnNumberStart": 1,
          "lineNumberStart": 3,
          "type": "LITERAL_NUMBER",
          "value": "6",
        },
        "lineNumberStart": 3,
        "operator": "^",
        "right": {
          "columnNumberStart": 5,
          "lineNumberStart": 3,
          "type": "LITERAL_NUMBER",
          "value": "2",
        },
        "type": "BINARY_EXPR",
      },
      "lineNumberStart": 3,
      "operator": "*",
      "right": {
        "columnNumberStart": 9,
        "lineNumberStart": 3,
        "type": "LITERAL_NUMBER",
        "value": "3",
      },
      "type": "BINARY_EXPR",
    },
    {
      "columnNumberStart": 3,
      "left": {
        "columnNumberStart": 1,
        "lineNumberStart": 4,
        "type": "IDENTIFIER",
        "value": "x",
      },
      "lineNumberStart": 4,
      "operator": "->",
      "right": {
        "children": [
          {
            "columnNumberStart": 8,
            "lineNumberStart": 4,
            "type": "IDENTIFIER",
            "value": "z",
          },
        ],
        "columnNumberStart": 6,
        "function": {
          "columnNumberStart": 6,
          "lineNumberStart": 4,
          "type": "IDENTIFIER",
          "value": "y",
        },
        "lineNumberStart": 4,
        "type": "FUNCTION_CALL",
      },
      "type": "BINARY_EXPR",
    },
    {
      "children": [
        {
          "columnNumberStart": 17,
          "left": {
            "columnNumberStart": 12,
            "left": {
              "columnNumberStart": 7,
              "left": {
                "columnNumberStart": 5,
                "lineNumberStart": 5,
                "type": "IDENTIFIER",
                "value": "b",
              },
              "lineNumberStart": 5,
              "operator": "==",
              "right": {
                "columnNumberStart": 10,
                "lineNumberStart": 5,
                "type": "IDENTIFIER",
                "value": "c",
              },
              "type": "BINARY_EXPR",
            },
            "lineNumberStart": 5,
            "operator": "||",
            "right": {
              "columnNumberStart": 15,
              "lineNumberStart": 5,
              "type": "IDENTIFIER",
              "value": "d",
            },
            "type": "BINARY_EXPR",
          },
          "lineNumberStart": 5,
          "operator": "!=",
          "right": {
            "columnNumberStart": 20,
            "lineNumberStart": 5,
            "type": "IDENTIFIER",
            "value": "e",
          },
          "type": "BINARY_EXPR",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 5,
      "type": "ASSIGNMENT",
      "variable": "a",
    },
    {
      "children": [
        {
          "columnNumberStart": 36,
          "left": {
            "columnNumberStart": 31,
            "left": {
              "columnNumberStart": 27,
              "left": {
                "columnNumberStart": 22,
                "left": {
                  "columnNumberStart": 17,
                  "left": {
                    "columnNumberStart": 12,
                    "left": {
                      "columnNumberStart": 7,
                      "left": {
                        "columnNumberStart": 5,
                        "lineNumberStart": 6,
                        "type": "IDENTIFIER",
                        "value": "g",
                      },
                      "lineNumberStart": 6,
                      "operator": ">=",
                      "right": {
                        "columnNumberStart": 10,
                        "lineNumberStart": 6,
                        "type": "IDENTIFIER",
                        "value": "h",
                      },
                      "type": "BINARY_EXPR",
                    },
                    "lineNumberStart": 6,
                    "operator": "&&",
                    "right": {
                      "columnNumberStart": 15,
                      "lineNumberStart": 6,
                      "type": "IDENTIFIER",
                      "value": "i",
                    },
                    "type": "BINARY_EXPR",
                  },
                  "lineNumberStart": 6,
                  "operator": "<=",
                  "right": {
                    "columnNumberStart": 20,
                    "lineNumberStart": 6,
                    "type": "IDENTIFIER",
                    "value": "j",
                  },
                  "type": "BINARY_EXPR",
                },
                "lineNumberStart": 6,
                "operator": "||",
                "right": {
                  "columnNumberStart": 25,
                  "lineNumberStart": 6,
                  "type": "IDENTIFIER",
                  "value": "k",
                },
                "type": "BINARY_EXPR",
              },
              "lineNumberStart": 6,
              "operator": ">",
              "right": {
                "columnNumberStart": 29,
                "lineNumberStart": 6,
                "type": "IDENTIFIER",
                "value": "l",
              },
              "type": "BINARY_EXPR",
            },
            "lineNumberStart": 6,
            "operator": "&&",
            "right": {
              "columnNumberStart": 34,
              "lineNumberStart": 6,
              "type": "IDENTIFIER",
              "value": "m",
            },
            "type": "BINARY_EXPR",
          },
          "lineNumberStart": 6,
          "operator": "<",
          "right": {
            "columnNumberStart": 38,
            "lineNumberStart": 6,
            "type": "IDENTIFIER",
            "value": "n",
          },
          "type": "BINARY_EXPR",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 6,
      "type": "ASSIGNMENT",
      "variable": "f",
    },
    {
      "children": [
        {
          "columnNumberStart": 12,
          "left": {
            "columnNumberStart": 9,
            "left": {
              "columnNumberStart": 6,
              "left": {
                "columnNumberStart": 5,
                "lineNumberStart": 7,
                "type": "IDENTIFIER",
                "value": "p",
              },
              "lineNumberStart": 7,
              "operator": ".",
              "right": {
                "columnNumberStart": 7,
                "lineNumberStart": 7,
                "type": "IDENTIFIER",
                "value": "q",
              },
              "type": "BINARY_EXPR",
            },
            "lineNumberStart": 7,
            "operator": "*",
            "right": {
              "columnNumberStart": 11,
              "lineNumberStart": 7,
              "type": "IDENTIFIER",
              "value": "r",
            },
            "type": "BINARY_EXPR",
          },
          "lineNumberStart": 7,
          "operator": ".",
          "right": {
            "columnNumberStart": 13,
            "lineNumberStart": 7,
            "type": "IDENTIFIER",
            "value": "s",
          },
          "type": "BINARY_EXPR",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 7,
      "type": "ASSIGNMENT",
      "variable": "o",
    },
  ],
  "type": "ROOT",
}
`;

exports[`ðŸ“¸ Operator precedence 3`] = `"2*3+4/5;( -10)+3%5;Math.pow(6,2)*3;y(x,z);const a=b===cundefinedd!==e;const f=g>=hundefinedi<=jundefinedk>lundefinedm<n;const o=p.q*r.s"`;
